### ISSUE TITLE
Blanket disable torch function/dispatch mode/subclass in dynamo

### ISSUE THREAD

**Issue by ezyang** [pytorch/torchdynamo#1822]
I noticed in https://github.com/pytorch/pytorch/pull/88218 that there are a lot of one off DisableTorchFunction all over the torchdynamo codebase. This seems unsustainable. We ought to have some clear invariant about how dynamically scoped state and subclasses affect dynamo. What is this invariant? The obvious one to me is that all modes / context managers are disabled in the compiler, and you explicitly restore the modes if you want to evaluate through them. Is this what we are doing?

**Comment by ezyang** [comment]
Cc @samdow @albanD @zou3519 @voznesenskym

**Comment by zou3519** [comment]
For now I would hope that dynamo falls back to eager mode when it sees a Tensor subclass (https://github.com/pytorch/torchdynamo/issues/42) or a mode or a context manager.

> The obvious one to me is that all modes / context managers are disabled in the compiler, and you explicitly restore the modes if you want to evaluate through them.

We've talked about this before, or is this something different? The previous thought was that dynamo would capture a graph with the enter/exit (not necessarily in the same graph!) and pass it to AOTDispatcher which would actually restore the modes/context managers and trace through them.

**Comment by ezyang** [comment]
Yeah, I think this is what we talked about before. In particular, when running dynamo itself, the modes should not be on, until we get to aot dispatcher.
